File: 2021/10/19/RangeSumBST.java
/**
 * https://leetcode.com/problems/range-sum-of-bst/
 * @author Andrew Jarombek
 * @since 10/19/2021
 */

public class RangeSumBST {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() { }
        TreeNode(int val) {
            this.val = val;
        }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    public int rangeSumBST(TreeNode root, int low, int high) {
        int sum = 0;

        if (low <= root.val && root.val <= high) {
            sum += root.val;
        }

        if (root.val >= low && root.left != null) {
            sum += rangeSumBST(root.left, low, high);
        }

        if (root.val <= high && root.right != null) {
            sum += rangeSumBST(root.right, low, high);
        }

        return sum;
    }
}

File: 2022/05/25/rangeSumBST.ts
/**
 * https://leetcode.com/problems/range-sum-of-bst/
 * @author Andrew Jarombek
 * @since 5/25/2022
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

function rangeSumBST(root: TreeNode | null, low: number, high: number): number {
    return dfs(root, low, high, 0);
}

function dfs(root: TreeNode | null, low: number, high: number, total: number): number {
    if (!root) {
        return total;
    }

    total = dfs(root.left, low, high, total);

    if (root.val >= low && root.val <= high) {
        total += root.val;
    }

    total = dfs(root.right, low, high, total);

    return total;
}